packages/wrangler/src/config/validation-helpers.ts

```diff
@@ -225,6 +225,30 @@ export const isString: ValidatorFn = (diagnostics, field, value) => {
   return true;
 };
 
+/**
+ * Validate that the `name` field is compliant with EWC constraints.
+ */
+export const isValidName: ValidatorFn = (diagnostics, field, value) => {
+  // Predicate typeguard returns `string` to previously `unknown` variable
+  function isDefinedString(input: unknown): input is string {
+    return input !== undefined && typeof input === "string";
+  }
+
+  if (
+    (isDefinedString(value) && /^$|^[a-z0-9_ ][a-z0-9-_ ]*$/.test(value)) ||
```

Comment:
We should throw an error if the name includes `_` and has `workers_dev` to (implicitly or explicitly) true.

This is because we cannot generate `workers.dev` SSL certificates with an `_` in the hostname.

Reply:
Interesting, this was the RegEx from the Go backend... Should I talk to Aaron or Josh about it again?

Reply:
You know what I think you're right, it's best just to remove the underscore altogether.

Reply:
Removing `_` eligibility would be a breaking change and prevent people from using Wrangler 2 for Workers with underscores.

As I understand it, the only constraint is that we cannot use underscores with `workers.dev` subdomains. So I propose that we do allow `_` in the Worker name, but do not attempt to publish to `workers.dev` if there is an underscore in the name, displaying a suitable warning when that happens.

Reply:
What we can do is during `wrangler publish` if the `workers_dev` is true and the name has underscores we can prevent the publish with a message.